### ISSUE TITLE
vimscript wrong indentation

### ISSUE THREAD

**Issue by habamax** [vim/vim#15123]
### Steps to reproduce

1. `vim -Nu NONE`
2. `:set ft=vim`
3. add following 2 lines
```
set path=.,,
set clipboard=unnamed,unnamedplus
```
4. goto line 2 and press `==`

### Expected behaviour

Line 2 shouldn't be indented.
![image](https://github.com/vim/vim/assets/234774/c926f2c5-b8fb-4456-8fce-741167c1fb1c)

### Version of Vim

9.1.516

### Environment

debian 12, bash

**Comment by a2chang** [comment]
Using AI by augmentcode.com to see what's going on:


You
> In this text, why does vim indent the second line?
> set path-.,,
>     set clipboard=unnamed,unnamedplus

Augment
> In the text you provided:
```
set path-.,,
    set clipboard=unnamed,unnamedplus
```
> Vim indents the second line because it's likely interpreting this as a continuation of the first command. This behavior is related to Vim's syntax highlighting and automatic indentation features for Vim script.
>
> Here's why this happens:
> 1. The first line set path-.,, ends with a comma. In many programming languages and in Vim script, a trailing comma can indicate that the statement continues on the next line.
> 1. Vim's built-in syntax highlighting for Vim script recognizes this pattern and assumes that the next line is a continuation of the set command.
> 1. As a result, Vim automatically indents the second line to visually represent this continuation.
>
> However, in this specific case, the indentation is actually incorrect because these are two separate set commands. They should be on separate lines without indentation:
```
set path-.,,
set clipboard=unnamed,unnamedplus
```

**Comment by a2chang** [comment]
You

> How can I modify vim to fix this?

Augment

> To fix this indentation issue in Vim, you have a few options:
>
> 1. Modify the indentation rules for Vim script: You can adjust the indent file for Vim script. This file is typically located at runtime/indent/vim.vim. In this file, you could add logic to prevent indentation after a line ending with a comma when the next line starts with 'set'.
> 1. Use autocmd to correct the indentation: Add the following to your vimrc file:
> `autocmd FileType vim autocmd BufWritePre <buffer> silent! :g/^\s*set/s/^\s\+//`
> This will remove leading whitespace from lines starting with 'set' when you save a Vim script file.
> 1. Modify the syntax highlighting: You could adjust the syntax file for Vim script (usually runtime/syntax/vim.vim) to not treat a comma at the end of a 'set' command as a line continuation.
> 1. Create a custom indentation function: In your vimrc, you could define a custom indentation function for Vim script files:
> `function! GetVimIndent()`
> `  let lnum = prevnonblank(v:lnum - 1)`
> `  if getline(v:lnum) =~ '^\s*set\>'`
> `    return 0`
> `  endif`
> `  return vim#indent#GetVimIndent()`
> `endfunction`
> ` `
> `autocmd FileType vim setlocal indentexpr=GetVimIndent()`
> This function checks if the current line starts with 'set' and if so, sets the indent to 0. Otherwise, it uses the default Vim script indentation.
> 1. Use a plugin: There might be Vim plugins available that provide more sophisticated indentation rules for Vim script.
>
> The most robust solution would be to modify the built-in indent file, but this requires careful consideration to avoid unintended consequences. The autocmd or custom indentation function approaches are safer as they only affect your personal Vim setup.
> Remember to test any changes thoroughly to ensure they don't cause issues with other valid Vim script constructs.

**Comment by a2chang** [comment]
The AI suggests this change to runtime/indent/vim.vim:

```
vim9script

# Vim indent file
# Language:	Vim script
# Maintainer:	The Vim Project <https://github.com/vim/vim>
# Last Change:	2024 Jul 12
# Former Maintainer:	Bram Moolenaar <Bram@vim.org>

# Only load this indent file when no other was loaded.
if exists('b:did_indent')
    finish
endif

b:did_indent = true
b:undo_indent = 'setlocal indentkeys< indentexpr<'

import autoload '../autoload/dist/vimindent.vim'

# export indentexpr as a global function, so it can be easily manipulated by plugins
g:VimIndent = vimindent.Expr

def CustomVimIndent(): number
    if getline(v:lnum) =~ '^\s*set\>'
        return 0
    endif
    return g:VimIndent()
enddef

setlocal indentexpr=CustomVimIndent()
setlocal indentkeys+==endif,=enddef,=endfu,=endfor,=endwh,=endtry,=endclass,=endinterface,=endenum,=},=else,=cat,=finall,=END,0\\
execute('setlocal indentkeys+=0=\"\\\ ,0=#\\\ ')
setlocal indentkeys-=0#
setlocal indentkeys-=:
```

Happy to make a PR if this is a reasonable change - I'm not literate in .vim files.

**Comment by jmdevin** [comment]
I don't think that making multiple comments to copy-paste answers which you don't have the expertise to review, and which are generated by an AI - made by a company you seem to work for - is an effective method of contributing to this project. You might be posting code that doesn't even work and asking others to look into it for you.

Instead, perhaps you could make the change to the indent script yourself and test them? If they work, then a pull request might be appropriate. Reviewing code that is at least attested to run without throwing errors is a better use of others' time.

This isn't meant to be a disparaging comment, but a practical one, and if the maintainers disagree with me then I'm in the wrong.

**Comment by habamax** [comment]
> Happy to make a PR if this is a reasonable change - I'm not literate in .vim files

Solution of AI is wrong:

1. there is also `setlocal` and `setglobal`
2. set commands might be deep in nested vimscript (if/else/while/def/lambdas etc)
3. fix should be targeting `/autoload/dist/vimindent.vim`

**Comment by chrisbra** [comment]
It seems we can fix this, if we add `:setglobal/setlocal` to `TRICKY_COMMANDS`. This assumes, `:set` with arguments basically never starts a line-continuation.
```diff
diff --git a/runtime/autoload/dist/vimindent.vim b/runtime/autoload/dist/vimindent.vim
index 07b21538e..3ea15bbe5 100644
--- a/runtime/autoload/dist/vimindent.vim
+++ b/runtime/autoload/dist/vimindent.vim
@@ -346,6 +346,7 @@ patterns =<< trim eval END
     \<[lt]\=cd!\=\s\+-\s*$
     \<norm\%[al]!\=\s*\S\+$
     \%(\<sil\%[ent]!\=\s\+\)\=\<[nvxsoilct]\=\%(nore\|un\)map!\=\s
+    \%(set\%(\%[global]\|\%[local]\)\).*,$
     {PLUS_MINUS_COMMAND}
 END

```

ping @lacygoill

**Comment by a2chang** [comment]
> I don't think that making multiple comments to copy-paste answers which you don't have the expertise to review, and which are generated by an AI - made by a company you seem to work for - is an effective method of contributing to this project. You might be posting code that doesn't even work and asking others to look into it for you.
> 
> Instead, perhaps you could make the change to the indent script yourself and test them? If they work, then a pull request might be appropriate. Reviewing code that is at least attested to run without throwing errors is a better use of others' time.
> 
> This isn't meant to be a disparaging comment, but a practical one, and if the maintainers disagree with me then I'm in the wrong.

No offence taken.  I was testing out our product, and have changes that work.  (Sorry I should have been more explicit about that.  Nobody wants random code that hasn't been tested.)  But that doesn't mean it's a reasonable chance, so I was looking for opinions before just making a PR.

Having said that, I did test it and it solves that issue and passes the tests.  I can add a test for this case and make a PR if you would like.
